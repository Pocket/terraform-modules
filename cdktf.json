{
  "language": "typescript",
  "app": "npm run --silent compile && node dist/src/example.js",
  "terraformProviders": [
    "pagerduty/pagerduty@~> 1.8.0",
    "archive@~> 1.3.0",
    "local@~> 2.1.0"
  ]
}
