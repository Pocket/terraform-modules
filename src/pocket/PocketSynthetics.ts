import { Resource } from 'cdktf';
import { Construct } from 'constructs';

import {
  SyntheticsMonitor,
  NrqlAlertCondition,
} from '@cdktf/provider-newrelic';

export interface PocketSyntheticProps {
  uri: string;
  verifySsl: boolean;
  policyId?: number;
  nrqlConfig?: {
    query?: string;
    evaluationOffset?: number;
    valueFunction?: string;
    violationTimeLimitSeconds?: number;
    closeViolationsOnExpiration?: boolean;
    expirationDuration?: number;
    critical?: {
      operator?: string;
      threshold?: number;
      thresholdDuration?: number;
      thresholdOccurrences?: string;
    };
  };
}

const globalCheckLocations = [
  'AWS_US_WEST_2',
  'AWS_EU_WEST_2',
  'AWS_US_EAST_2',
];

export class PocketSyntheticCheck extends Resource {
  constructor(
    scope: Construct,
    private name: string,
    private config: PocketSyntheticProps
  ) {
    super(scope, name);

    // if policy id not provided use default policy id
    // policy in another
    if (this.config.policyId === undefined) {
      // I wanted to do a terraform lookup but it doesn't run before generating
      // the cdktf.json file so we have to hardcode the default policy id
      this.config.policyId = 1707149; // Pocket-Default-Policy
    }

    const pocketMonitor = new SyntheticsMonitor(
      this,
      `${this.name}-synthetics-monitor`,
      {
        name: `${this.name}-synthetics`,
        type: 'SIMPLE',
        frequency: 5,
        status: 'ENABLED',
        locations: globalCheckLocations,
        uri: this.config.uri,
        verifySsl: this.config.verifySsl,
      }
    );

    const defaultNrqlConfig = {
      query: `SELECT count(result) from SyntheticCheck where result = 'FAILED' and monitorName = '${pocketMonitor.name}'`,
      evaluationOffset: 3,
      valueFunction: 'sum',
      violationTimeLimitSeconds: 2592000,
      closeViolationsOnExpiration: true,
      expirationDuration: 600,
      critical: {
        operator: 'above',
        threshold: 2,
        thresholdDuration: 900,
        thresholdOccurrences: 'AT_LEAST_ONCE',
      },
    };

    this.config.nrqlConfig = {
      ...defaultNrqlConfig,
      ...config.nrqlConfig,
    };

    new NrqlAlertCondition(this, 'alert-condition', {
      name: `${this.name}-nrql`,
      policyId: this.config.policyId,
      fillValue: 0,
      fillOption: 'static',
      nrql: {
        query: this.config.nrqlConfig.query,
        evaluationOffset: this.config.nrqlConfig.evaluationOffset,
      },
      valueFunction: this.config.nrqlConfig.valueFunction,
      violationTimeLimitSeconds:
        this.config.nrqlConfig.violationTimeLimitSeconds,
      closeViolationsOnExpiration:
        this.config.nrqlConfig.closeViolationsOnExpiration,
      expirationDuration: this.config.nrqlConfig.expirationDuration,
      critical: {
        operator: this.config.nrqlConfig.critical.operator,
        threshold: this.config.nrqlConfig.critical.threshold,
        thresholdDuration: this.config.nrqlConfig.critical.thresholdDuration,
        thresholdOccurrences:
          this.config.nrqlConfig.critical.thresholdOccurrences,
      },
    });
  }
}
