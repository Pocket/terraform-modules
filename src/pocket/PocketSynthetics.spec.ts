import { Testing } from 'cdktf';
import { PocketSyntheticProps, PocketSyntheticCheck } from './PocketSynthetics';

const config: PocketSyntheticProps = {
  uri: 'acme.getpocket.dev',
  verifySsl: true,
};

it('renders a Pocket New Relic synthetic check', () => {
  const synthed = Testing.synthScope((stack) => {
    new PocketSyntheticCheck(stack, 'test-synthetic', config);
  });
  expect(synthed).toMatchSnapshot();
});

it('allows passing different values for nrql config', () => {
  config.nrqlConfig.query = 'SELECT * FROM MY-COOL-TABLE';
  const synthed = Testing.synthScope((stack) => {
    new PocketSyntheticCheck(stack, 'test-synthetic', config);
  });
  expect(synthed).toMatchSnapshot();
});
