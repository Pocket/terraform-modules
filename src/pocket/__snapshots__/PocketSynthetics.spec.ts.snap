// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`allows passing different values for nrql config 1`] = `
"{
  "resource": {
    "newrelic_nrql_alert_condition": {
      "test-synthetic_alert-condition_4C991B53": {
        "aggregation_delay": "180",
        "aggregation_method": "cadence",
        "aggregation_window": 3000,
        "close_violations_on_expiration": true,
        "critical": {
          "operator": "above",
          "threshold": 2,
          "threshold_duration": 900,
          "threshold_occurrences": "AT_LEAST_ONCE"
        },
        "expiration_duration": 600,
        "fill_option": "static",
        "fill_value": 0,
        "name": "test-synthetic-nrql",
        "nrql": {
          "query": "SELECT * FROM MY-COOL-TABLE"
        },
        "policy_id": 1707149,
        "slide_by": 60,
        "violation_time_limit_seconds": 2592000
      }
    },
    "newrelic_synthetics_monitor": {
      "test-synthetic_test-synthetic-synthetics-monitor_914DC880": {
        "locations_public": [
          "AWS_US_WEST_2",
          "AWS_EU_WEST_2",
          "AWS_US_EAST_2"
        ],
        "name": "test-synthetic-synthetics",
        "period": "EVERY_5_MINUTES",
        "status": "ENABLED",
        "type": "SIMPLE",
        "uri": "acme.getpocket.dev",
        "verify_ssl": true
      }
    }
  }
}"
`;

exports[`renders a Pocket New Relic synthetic check 1`] = `
"{
  "resource": {
    "newrelic_nrql_alert_condition": {
      "test-synthetic_alert-condition_4C991B53": {
        "aggregation_delay": "180",
        "aggregation_method": "cadence",
        "aggregation_window": 3000,
        "close_violations_on_expiration": true,
        "critical": {
          "operator": "above",
          "threshold": 2,
          "threshold_duration": 900,
          "threshold_occurrences": "AT_LEAST_ONCE"
        },
        "expiration_duration": 600,
        "fill_option": "static",
        "fill_value": 0,
        "name": "test-synthetic-nrql",
        "nrql": {
          "query": "SELECT count(result) from SyntheticCheck where result = 'FAILED' and monitorName = '\${newrelic_synthetics_monitor.test-synthetic_test-synthetic-synthetics-monitor_914DC880.name}'"
        },
        "policy_id": 1707149,
        "slide_by": 60,
        "violation_time_limit_seconds": 2592000
      }
    },
    "newrelic_synthetics_monitor": {
      "test-synthetic_test-synthetic-synthetics-monitor_914DC880": {
        "locations_public": [
          "AWS_US_WEST_2",
          "AWS_EU_WEST_2",
          "AWS_US_EAST_2"
        ],
        "name": "test-synthetic-synthetics",
        "period": "EVERY_5_MINUTES",
        "status": "ENABLED",
        "type": "SIMPLE",
        "uri": "acme.getpocket.dev",
        "verify_ssl": true
      }
    }
  }
}"
`;
